### DISCUSSION TITLE
Immutable F# Records and Updating Entities

### DISCUSSION THREAD

**Discussion by no1melman** [ravendb/ravendb#14930] [Q&A] [answered]
So if I have this

```fsharp

type MyEntity = 
  {
    Id: string;
    Associates: string list
  }

use session = client.OpenAsyncSession()

let! theentity = session.LoadAsync<MyEntity>("123")

// update associates
let updatedEntity = { theentity with Associates = "callum" :: theentity.Associates }

do! session.SaveChangesAsync()

```

This won't work will it? Because the reference has now changed... is this correct? if so, how does one make this work?

**Comment by ayende** [comment, marked as answer]
This goes directly against the design of RavenDB's session as a unit of work.
Working with immutable data doesn't make much sense in that context.

You _can_ make it work with explicit instructions to RavenDB, you'll need to do something like:

```
session.Advanced.Evict(original);
session.Store(updated);
```

**Reply by no1melman** [reply]
Any serious drawbacks to doing this? Or just can get sticky if doing a lot of it due to tracking changes?

**Reply by ayende** [reply]
The main reason I don't like it is that you are now basically moving all state management to yourself vs. letting RavenDB do that.

Put another way, you would probably be better off with this option, instead:

```
use readSession= client.OpenAsyncSession()
use writeSession= client.OpenAsyncSession()

// read from readSession
// use immutable records
// write to the session
```

Where it gets really tricky is when you have something like this:


```
var currentOrder = session.Load<Order>(orderId);
var recentOrders = session.Query<Order>().OrderByDescending(x=>x.Date).Take(5).ToList();

// mutate both currentOrder and recentOrders 

// now save it.
```

What is the problem? If `currentOrder` also includes `recentOrders`, you are losing the identity map, which will insure that you'll get the same _instance_ from the session. That means that you might get an error or strange results on persistance.
